### ISSUE TITLE
Elevation not converted to projected South Africa CS (promoted to 3D)

### ISSUE THREAD

**Issue by jjimenezshaw** [OSGeo/PROJ#2170]
This problem is very similar to #2122. This one is fixed and works in my checkout.

#### Example of problem
Converting from `EPSG:4979` to `EPSG:4222` (South African geographical crs, `2D`) does not compute the vertical difference. If I promote `EPSG:4222` to `3D`, it works. However, Converting from `EPSG:4979` to `EPSG:22275` (promoted to `3D`), that is a projected crs over `EPSG:4222`, does not change the Z value.

Notice that `EPSG:22275` projection is **`PROJECTION["Transverse_Mercator_South_Orientated"]`**

and some WKT has this parameter `TOWGS84[-134,-110,-292,0,0,0,0]`

#### Problem description
These commands promote to 3D `EPSG:22275` and `EPSG:4222` to `sa3dp` and `sa3dg`:
`sa3dp=$(projinfo EPSG:22275 --3d -o PROJJSON -q --single-line)`
`sa3dg=$(projinfo EPSG:4222 --3d -o PROJJSON -q --single-line)`

Using these promoted crs, we see how the geographic one compute the transformation in Z between these two ellipsoids
`echo -33.9189 18.4233 0 | cs2cs EPSG:4979 "$sa3dg" -f "%.6f" => -33.918809 18.423943 -31.116975`

while the projected one doesn't
`echo -33.9189 18.4233 0 | cs2cs EPSG:4979 "$sa3dp" -f "%.6f" => -316643.645813 3759668.647732 0.000000`

(See below debug outputs. Notice that projected one is much shorter than the geographic one)

Running with GDAL (that is how I noticed it), I get in the concole the error message `ERROR 1: PROJ: proj_as_wkt: Projected 3D CRS can only be exported since WKT2:2019`, but only with the Projected CRS, not with the Geographic one.

#### Expected Output
I was expecting the same Z computation between the two ellipsoids, independently of the projection or not.

I don't know if the issue is only with South African projection `PROJECTION["Transverse_Mercator_South_Orientated"]` or may affect other projections as well.

#### Environment Information
 - PROJ version (`proj`) `master` at `ea18e39b3b4`
 - Operation System Information: Ubuntu 18.04

#### Installation method
Compiled from source (master at ea18e39b3b4) with CMake.

#### Debug outputs
with `PROJ_DEBUG=3` I got these outputs.
Projected CRS:
```
pj_open_lib(proj.db): call fopen(/home/jshaw/work/proj/build/data/proj.db) - succeeded
pj_open_lib(proj.ini): call fopen(/home/jshaw/work/proj/build/data/proj.ini) - succeeded
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final:    
Pipeline: Building arg list for step no. 0
Pipeline: init - proj=axisswap, 2
    order=2,1
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Pipeline: Step 0 (proj=axisswap) at 0x5626bdf5e050
Pipeline at [0x5626be031810]:    step at [0x5626bdf5e050] (proj=axisswap) done
Pipeline: Building arg list for step no. 1
Pipeline: init - proj=unitconvert, 5
    xy_in=deg
    z_in=m
    xy_out=rad
    z_out=m
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
xy_in unit: Degree
xy_out unit: Radian
z_in unit: Meter
z_out unit: Meter
Pipeline: Step 1 (proj=unitconvert) at 0x5626bdcd7a40
Pipeline at [0x5626be031810]:    step at [0x5626bdcd7a40] (proj=unitconvert) done
Pipeline: Building arg list for step no. 2
Pipeline: init - proj=tmerc, 9
    axis=wsu
    lat_0=0
    lon_0=15
    k=1
    x_0=0
    y_0=0
    a=6378249.145
    rf=293.4663077
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Pipeline: Step 2 (proj=tmerc) at 0x5626be0367d0
Pipeline at [0x5626be031810]:    step at [0x5626be0367d0] (proj=tmerc) done
Pipeline: 3 steps built. Determining i/o characteristics
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
-316643.645813	3759668.647732 0.000000
```

Geographical CRS:
```
pj_open_lib(proj.db): call fopen(/home/jshaw/work/proj/build/data/proj.db) - succeeded
pj_open_lib(proj.ini): call fopen(/home/jshaw/work/proj/build/data/proj.ini) - succeeded
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  ellps=WGS84
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final:    
Pipeline: Building arg list for step no. 0
Pipeline: init - proj=axisswap, 2
    order=2,1
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Pipeline: Step 0 (proj=axisswap) at 0x55e533526520
Pipeline at [0x55e5335261f0]:    step at [0x55e533526520] (proj=axisswap) done
Pipeline: Building arg list for step no. 1
Pipeline: init - proj=unitconvert, 5
    xy_in=deg
    z_in=m
    xy_out=rad
    z_out=m
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
xy_in unit: Degree
xy_out unit: Radian
z_in unit: Meter
z_out unit: Meter
Pipeline: Step 1 (proj=unitconvert) at 0x55e5334c6b20
Pipeline at [0x55e5335261f0]:    step at [0x55e5334c6b20] (proj=unitconvert) done
Pipeline: Building arg list for step no. 2
Pipeline: init - proj=cart, 2
    ellps=WGS84
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=WGS84
Pipeline: Step 2 (proj=cart) at 0x55e5334c6e50
Pipeline at [0x55e5335261f0]:    step at [0x55e5334c6e50] (proj=cart) done
Pipeline: Building arg list for step no. 3
Pipeline: init - proj=helmert, 4
    x=136
    y=108
    z=292
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Helmert parameters:
x=  136.00000  y=  108.00000  z=  292.00000
rx=  0.00000  ry=  0.00000  rz=  0.00000
s=   0.00000  exact=0
dx=  0.00000  dy=  0.00000  dz=  0.00000
drx= 0.00000  dry= 0.00000  drz= 0.00000
ds=  0.00000  t_epoch= 0.00000
Pipeline: Step 3 (proj=helmert) at 0x55e5334450e0
Pipeline at [0x55e5335261f0]:    step at [0x55e5334450e0] (proj=helmert) done
Pipeline: Building arg list for step no. 4
Pipeline: init - inv, 4
    proj=cart
    a=6378249.145
    rf=293.4663077
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  
Pipeline: Step 4 (inv) at 0x55e533445470
Pipeline at [0x55e5335261f0]:    step at [0x55e533445470] (inv) done
Pipeline: Building arg list for step no. 5
Pipeline: init - proj=longlat, 3
    a=6378249.145
    rf=293.4663077
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  
Pipeline: Step 5 (proj=longlat) at 0x55e5334da000
Pipeline at [0x55e5335261f0]:    step at [0x55e5334da000] (proj=longlat) done
Pipeline: Building arg list for step no. 6
Pipeline: init - proj=unitconvert, 5
    xy_in=rad
    z_in=m
    xy_out=deg
    z_out=m
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
xy_in unit: Radian
xy_out unit: Degree
z_in unit: Meter
z_out unit: Meter
Pipeline: Step 6 (proj=unitconvert) at 0x55e5333fe5e0
Pipeline at [0x55e5335261f0]:    step at [0x55e5333fe5e0] (proj=unitconvert) done
Pipeline: Building arg list for step no. 7
Pipeline: init - proj=axisswap, 2
    order=2,1
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Pipeline: Step 7 (proj=axisswap) at 0x55e533442760
Pipeline at [0x55e5335261f0]:    step at [0x55e533442760] (proj=axisswap) done
Pipeline: 8 steps built. Determining i/o characteristics
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  ellps=WGS84
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final:    
Pipeline: Building arg list for step no. 0
Pipeline: init - proj=axisswap, 2
    order=2,1
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Pipeline: Step 0 (proj=axisswap) at 0x55e533442af0
Pipeline at [0x55e5333c93d0]:    step at [0x55e533442af0] (proj=axisswap) done
Pipeline: Building arg list for step no. 1
Pipeline: init - proj=unitconvert, 5
    xy_in=deg
    z_in=m
    xy_out=rad
    z_out=m
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
xy_in unit: Degree
xy_out unit: Radian
z_in unit: Meter
z_out unit: Meter
Pipeline: Step 1 (proj=unitconvert) at 0x55e5333c90a0
Pipeline at [0x55e5333c93d0]:    step at [0x55e5333c90a0] (proj=unitconvert) done
Pipeline: Building arg list for step no. 2
Pipeline: init - proj=cart, 2
    ellps=WGS84
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=WGS84
Pipeline: Step 2 (proj=cart) at 0x55e5333e6680
Pipeline at [0x55e5333c93d0]:    step at [0x55e5333e6680] (proj=cart) done
Pipeline: Building arg list for step no. 3
Pipeline: init - proj=helmert, 4
    x=136
    y=108
    z=292
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Helmert parameters:
x=  136.00000  y=  108.00000  z=  292.00000
rx=  0.00000  ry=  0.00000  rz=  0.00000
s=   0.00000  exact=0
dx=  0.00000  dy=  0.00000  dz=  0.00000
drx= 0.00000  dry= 0.00000  drz= 0.00000
ds=  0.00000  t_epoch= 0.00000
Pipeline: Step 3 (proj=helmert) at 0x55e5333f82c0
Pipeline at [0x55e5333c93d0]:    step at [0x55e5333f82c0] (proj=helmert) done
Pipeline: Building arg list for step no. 4
Pipeline: init - inv, 4
    proj=cart
    a=6378249.145
    rf=293.4663077
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  
Pipeline: Step 4 (inv) at 0x55e5333ebd60
Pipeline at [0x55e5333c93d0]:    step at [0x55e5333ebd60] (inv) done
Pipeline: Building arg list for step no. 5
Pipeline: init - proj=longlat, 3
    a=6378249.145
    rf=293.4663077
pj_ellipsoid - final: a=6378249.145 f=1/293.466, errno=0
pj_ellipsoid - final: a=6378249.145 rf=293.4663077  
Pipeline: Step 5 (proj=longlat) at 0x55e53348ae90
Pipeline at [0x55e5333c93d0]:    step at [0x55e53348ae90] (proj=longlat) done
Pipeline: Building arg list for step no. 6
Pipeline: init - proj=unitconvert, 5
    xy_in=rad
    z_in=m
    xy_out=deg
    z_out=m
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
xy_in unit: Radian
xy_out unit: Degree
z_in unit: Meter
z_out unit: Meter
Pipeline: Step 6 (proj=unitconvert) at 0x55e5333b56d0
Pipeline at [0x55e5333c93d0]:    step at [0x55e5333b56d0] (proj=unitconvert) done
Pipeline: Building arg list for step no. 7
Pipeline: init - proj=axisswap, 2
    order=2,1
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Pipeline: Step 7 (proj=axisswap) at 0x55e533459560
Pipeline at [0x55e5333c93d0]:    step at [0x55e533459560] (proj=axisswap) done
Pipeline: 8 steps built. Determining i/o characteristics
pj_ellipsoid - final: a=6378137.000 f=1/298.257, errno=0
pj_ellipsoid - final:    ellps=GRS80
Using coordinate operation Inverse of Cape to WGS 84 (1)
-33.918809	18.423943 -31.116975
```

**Comment by rouault** [comment]
Things work as expected.
The non-null transformation EPSG:1128 "Cape to WGS 84 (1)" between EPSG:4979 and EPSG:4222 has area of validity 
AREA["South Africa - onshore"],
BBOX[-34.88,16.45,-22.13,32.95]]

But EPSG:22275 has
AREA["Namibia - Walvis Bay"],
BBOX[-23.15,14.35,-22.68,14.6]

Consequently EPSG:1128 "Cape to WGS 84 (1)" cannot be applied to EPSG:22275 and a ballpart / null datum transformation is used.

**Comment by jjimenezshaw** [comment]
Thank you. Makes sense.

Just for curiosity I tried with `EPSG:22277`
`sa3dp2=$(./projinfo EPSG:22277 --3d -o PROJJSON -q --single-line)`
`echo -33.9189 18.4233 0 | cs2cs EPSG:4979 "$sa3dp2" -f "%.6f" => -131685.431201 3755289.486832 -31.116975`
it works.

However it has
AREA["South Africa - west of 18°E"],
BBOX[-33.1,16.45,-28.03,18]
that doesn't contain the input latitude -33.9189
Why?

**Comment by rouault** [comment]
> that doesn't contain the input latitude -33.9189

yeah, that's a bit complicated.

Here's there some overlap between AREA["South Africa - west of 18°E"],BBOX[-33.1,16.45,-28.03,18] and AREA["South Africa - onshore"],BBOX[-34.88,16.45,-22.13,32.95]], so this transformation is kept, and it can be applied tow whole BBOX[-34.88,16.45,-22.13,32.95]], into which  -33.9189 18.4233 falls. The area of use of the projection BBOX[-33.1,16.45,-28.03,18] doesn't influence the whole transformation, as a projection is something "exact" that can be used beyond its area of use.
And now you're confused.
Basically the way PROJ computes transformations is based on a set of (hopefully reasonable) hypothesis, and when 2 BBOX[] don't interesect at all, this is used to filter out apparently inconsistent results.

**Comment by jjimenezshaw** [comment]
oh, thanks.
Looks like EPSG:22275 is inconsistently defined. The "projected area of interest" does not overlap its own "geographic area of interest". The former is defined far in Namibia, while the later is South Africa.
